Avoid panic with recordsize > 128k, raw sending and no large_blocks

gamanakis opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://openzfs.github.io/openzfs-docs/Developer%20Resources/Buildbot%20Options.html
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
See #12275.

### Description
<!--- Describe your changes in detail -->
The current codebase does not support raw sending buffers with block
size > 128kB when large_blocks is not active. Instead of panicking
due to a NULL pointer dereference in dmu_dump_write() print out an
error message.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->
Running the replicator in #12275 does not panic.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

scineram commented:
> buffers with block size > 128kB when large_blocks is not active

At first I thought this was nonsense. Is the issue here sending large blocks raw but without -L flag? Then shouldn't it be implied like the -c flag?

gamanakis commented:
> 
> 
> > with block size > 128kB when large_blocks is not active
> 
> At first I thought this was nonsense. Is the issue here sending large blocks raw but without -L flag? Then shouldn't it be implied like the -c flag?

Right, sending raw should always imply large blocks if the recordsize is >128kB. However, issue #12275 describes a situation where raw sending panics because large_blocks is not active on the pool.

gamanakis commented:
I think there may be a better way to report the error.

gamanakis commented:
@pcd1193182 would you mind taking a look at this? Thank you.

gamanakis commented:
@amotin thanks for taking a look! In the current codebase I cannot see how the activation code in dsl_dataset_block_count() could be bypassed. I still believe though that a scrub should be checking for this possibility.

gamanakis commented:
I am putting this one on hold, till we have a better understanding of the underlying problem.

gamanakis commented:
302e0f3: rebased to master, and implemented a new logic: the current codebase does not support raw sending buffers with block size > 128kB when large_blocks is not active. This can happen in the codepath dsl_dataset_sync()->dmu_objset_sync()->zio_nowait() which calls back dmu_objset_write_done()->dsl_dataset_block_born().

If dsl_dataset_sync() completes its run before dsl_dataset_block_born() is called, we will end up not activating some of the necessary flags, while having blocks based on those flags written in the filesystem. A subsequent send will then panic.

Fix this by directly deciding in dmu_objset_sync() whether these flags need to be activated later by dsl_dataset_sync(). Also during scrub verify there are no contradicting filesystem flags.

jumbi77 commented:
Thanks for this fix! Just a (dump) question: Does ist make sense to write a test case to test if the necessary flags are implicitly activate when sending a snapshot?

gamanakis commented:
> Does ist make sense to write a test case to test if the necessary flags are implicitly activate when sending a snapshot?

That's a good question; the corner case this PR resolves is however difficult to trigger in a test.

gamanakis commented:
[51341c2](https://github.com/openzfs/zfs/commit/51341c225cd7b2b80b42c8e1724100d8b975ecb8): rebased to master.

gamanakis commented:
3dfb54d: Rebased to master.
